text/2021-04-23-builtin-gc-worker.md

```diff
@@ -0,0 +1,131 @@
+
+2. `Transaction Safepoint`
+
+    TiKV and TiFlash guarantee that no read or write will happen on the data earlier than `Transaction Safepoint`. PD calculates `Transaction Safepoint Proposal` by `min(GC Barriers) - 1` and proposes it to TiKV and TiFlash via the `StoreHeartbeatResponse`. Therefore, TiKV and TiFlash will reject all requests with timestamp earlier than the `Transaction Safepoint`. In the next following `StoreHeartbeatRequest` TiKV and TiFlash will report their latest `Transaction Safepoint Status` to PD.
+
+3. `GC Safepoint`
+
+    PD calculates `GC Safepoint` by `min(Transaction Safepoint Status) - 1`, resolves locks earlier than `GC Safepoint`, then synchronizes it to TiKV and TiFlash via `StoreHeartbeatResponse`. And finally, TiKV and TiFlash are safe to delete MVCC data earlier than `GC Safepoint`.
+
+![GC Worker](../media/gc-worker.png)
+
+According to the design above, several fields will be added to `pdpb.proto`:
+
+```diff
+  // `pdpb.proto
+
+  message StoreHeartbeatRequest {
+      RequestHeader header = 1;
+      StoreStats stats = 2;
+  }
+
+  message StoreStats {
+  …
++     uint64 txn_safepoint = 15;
++     uint64 gc_safepoint = 16;
+  }
+
+  message PutStoreResponse {
+      ResponseHeader header = 1;
+      replication_modepb.ReplicationStatus replication_status = 2;
++     uint64 txn_safepoint = 3;
++     uint64 gc_safepoint = 4;
+  }
+
+  message StoreHeartbeatResponse {
+     ResponseHeader header = 1;
+     replication_modepb.ReplicationStatus replication_status = 2;
+     string cluster_version = 3;
++    uint64 txn_safepoint = 4;
++    uint64 gc_safepoint = 5;
+  }
+```
+
+Aslo, RPC `SetGCBarrier` and `GetGCStatus` will be added:
+
+```diff
+  // `pdpb.proto
+
+  service PD {
+  ...
++    rpc SetGCBarrier(SetGCBarrierRequest) returns (SetGCBarrierResponse) {}
++    rpc GetGCStatus(GetGCStatusRequest) returns (GetGCStatusResponse) {}
+  }
+
++ message GCBarrier {
++    bytes id = 1;
++    // Set to 0 to delete existing barrier
++    uint64 ttl = 2;
++    uint64 timestamp = 3;
++ }
+
++ message SetGCBarrierRequest {
++    RequestHeader header = 1;
++    GCBarrier gc_barrier = 2;
++ }
+
++ message SetGCBarrierResponse {
++    ResponseHeader header = 1;
++ }
+
++ message GetGCStatusRequest {
++    RequestHeader header = 1;
++ }
+
++ message GetGCStatusResponse {
++    ResponseHeader header = 1;
++    repeated GCBarriers gc_barriers = 2;
++    uint64 txn_safepoint = 3;
++    uint64 gc_safepoint = 4;
++ }
+```
+
+### Backward compatibility
+
+#### Rolling update
+
+During the rolling update, the GC worker will not function and the new TiDB will not try to be elected as a GC leader. Once the rolling update is completed, the GC worker in PD will start to work.
+
+The RPC `GetGCSafePoint` and `UpdateServiceGCSafePoint`, which are used by the GC model before this proposal should still work for keep compatible with tools and they should be just a thin wrapper on top of the new `GetGCStatus` and `SetGCBarrier`.
```

Comment:
You also need to ensure that service safepoints that are already set in old version won't be lost and will become the GC barrier after upgrading.

Reply:
Yes, that should not be difficult, but not possible to downgrade in place. I've to solution for that: 1. reuse the underling service safepoint storage for GC barrier 2. always store both service safepoint and GC barrier.

Reply:
I think reusing the service safepoint storage would be easier. If you want to write both the service safepoint keys and your new keys at the same time, you need to consider how to keep them consistent when PD leader is transferring, or when PD is rolling upgrading (which means there are old version PD nodes).